src/gtk/window.cpp

```diff
@@ -2828,6 +2893,661 @@ static gboolean source_dispatch(GSource*, GSourceFunc, void*)
+                          G_CALLBACK(zoom_gesture_end_callback), win);
+    }
+    else
+    {
+        m_zoom_gesture = NULL;
+    }
+
+    if ( eventsMask & wxTOUCH_ROTATE_GESTURE )
+    {
+        eventsMask &= ~wxTOUCH_ROTATE_GESTURE;
+
+        m_rotate_gesture = gtk_gesture_rotate_new(widget);
+
+        gtk_event_controller_set_propagation_phase (GTK_EVENT_CONTROLLER(m_rotate_gesture), GTK_PHASE_TARGET);
+
+        g_signal_connect (m_rotate_gesture, "begin",
+                          G_CALLBACK(rotate_gesture_begin_callback), win);
+        g_signal_connect (m_rotate_gesture, "angle-changed",
+                          G_CALLBACK(rotate_gesture_callback), win);
+        g_signal_connect (m_rotate_gesture, "end",
+                          G_CALLBACK(rotate_gesture_end_callback), win);
+        g_signal_connect (m_rotate_gesture, "cancel",
+                          G_CALLBACK(rotate_gesture_end_callback), win);
+    }
+    else
+    {
+        m_rotate_gesture = NULL;
+    }
+
+    if ( eventsMask & wxTOUCH_PRESS_GESTURES )
+    {
+        eventsMask &= ~wxTOUCH_PRESS_GESTURES;
+
+        m_long_press_gesture = gtk_gesture_long_press_new(widget);
+
+        // "pressed" signal is also generated when left mouse is down for some minimum duration of time.
+        // This can be disable by calling gtk_gesture_single_set_touch_only(GTK_GESTURE_SINGLE(m_long_press_gesture), TRUE)
+        // which will allow "pressed" signal only for Touch events.
+
+        gtk_event_controller_set_propagation_phase(GTK_EVENT_CONTROLLER(m_long_press_gesture), GTK_PHASE_TARGET);
+
+        g_signal_connect (m_long_press_gesture, "pressed",
+                          G_CALLBACK(long_press_gesture_callback), win);
+    }
+    else
+    {
+        m_long_press_gesture = NULL;
+    }
+
+    wxASSERT_MSG( eventsMask == 0, "Unknown touch event mask bit specified" );
+
+    g_signal_connect (widget, "touch-event",
+                      G_CALLBACK(touch_callback), win);
+}
+
+wxWindowGesturesData::~wxWindowGesturesData()
+{
+    if ( m_vertical_pan_gesture )
+        g_object_unref(m_vertical_pan_gesture);
+    if ( m_horizontal_pan_gesture )
+        g_object_unref(m_horizontal_pan_gesture);
+    if ( m_zoom_gesture )
+        g_object_unref(m_zoom_gesture);
+    if ( m_rotate_gesture )
+        g_object_unref(m_rotate_gesture);
+    if ( m_long_press_gesture )
+        g_object_unref(m_long_press_gesture);
+}
+
+#endif // wxGTK_HAS_GESTURES_SUPPORT
+
+// This method must be always defined for GTK+ 3 as it's declared in the
+// header, where we can't (easily) test for wxGTK_HAS_GESTURES_SUPPORT.
+#ifdef __WXGTK3__
+
+bool wxWindowGTK::EnableTouchEvents(int eventsMask)
+{
+#ifdef wxGTK_HAS_GESTURES_SUPPORT
+    // Check if gestures support is also available during run-time.
+    if ( gtk_check_version(3, 14, 0) == NULL )
+    {
+        if ( eventsMask == wxTOUCH_NONE )
+        {
+            wxWindowGestures::EraseForObject(this);
```

Comment:
This will prevent disabling touch events from the event handler.

Reply:
Sorry, why?

Reply:
touch_callback ->  wxWindow::HandleWindowEvent -> SafelyProcessEvent -> ProcesEvent -> User-Event-Handler -> EnableTouchEvents(0);

touch_callback still uses data after returning from calling the event handler.

Reply:
Where does it use data after calling the event handler? I just don't see it...